doc/internals/icicle-v0.bnf

```diff
@@ -0,0 +1,304 @@
+#
+# Lexical Syntax
+#
+
+<program>      ::= <program-part> <program>
+                |  ε
+
+<program-part> ::= <lexeme>
+                |  <white-space>
+
+<lexeme>       ::= <literal>
+                |  <var-id>
+                |  <var-op>
+                |  <con-id>
+                |  <reserved-id>
+                |  <reserved-op>
+                |  <parens>
+
+<literal>      ::= <integer>
+                |  <float>
+                |  <string>
+                |  <date>
+
+<white-space>  ::= <white-space> <white-stuff>
+                |  ε
+<white-stuff>  ::= <white-char> | <comment>
+
+<white-char>   ::= ' ' | '\t' | '\n' | '\v'
+<white-chars>  ::= <white-chars> <white-char>
+                |  ε
+
+<comment>      ::= # <not-newline> <newline>
+
+<not-newline>  ::= [^\n]
+<not-newlines> ::= <not-newlines> <not-newline>
+                |  ε
+<newline>      ::= '\n'
+
+<lower>        ::= [a-z]
+<upper>        ::= [A-Z]
+
+<octit>        ::= [0-7]
+<digit>        ::= [0-9]
+<hexit>        ::= [0-9] | [a-f] | [A-F]
+
+<octal>        ::= <octal> <octit>
+                |  <octit>
+<decimal>      ::= <decimal> <digit>
+                |  <digit>
+<hexadecimal>  ::= <hexadecimal> <hexit>
+                |  <hexit>
+
+<integer>      ::= <decimal>
+                |  0o <octal>
+                |  0O <octal>
+                |  0x <hexadecimal>
+                |  0X <hexadecimal>
+
+<float>        ::= <decimal> . <decimal>
+                |  <decimal> . <decimal> <exponent>
+                |  <decimal> <exponent>
+
+<exponent>     ::= e <sign-opt> <decimal>
+                |  E <sign-opt> <decimal>
+
+<sign-opt>     ::= <sign>
+                |  ε
+<sign>         ::= -
+                |  +
+
+<projection>   ::= . <var-id>
+<projections>  ::= <projections> <projection>
+                |  ε
+
+<var-id>       ::= <var-head> <var-tails>
+<var-head>     ::= <lower>
+                |  _
+<var-tail>     ::= <lower>
+                |  <upper>
+                |  <digit>
+                |  _
+                |  $
+                |  '
+<var-tails>    ::= <var-tails> <var-tail>
+                |  ε
+
+<con-id>       ::= <con-head> <con-tails>
+<con-head>     ::= <upper>
+                |  _
```

Comment:
I don't think constructors can start with underscore, and I guess single "_" case is just a var